### ISSUE TITLE
ReflectMetadataProvider when using TS entity in JS API does not work

### ISSUE THREAD

**Issue by jonahallibone** [mikro-orm/mikro-orm#879]
**Describe the bug**
When defining entities in TS and then trying to initialize the ORM in a JS file, it seems as if the metadata for the entities is not discovered.

**Stack trace**
```
[discovery] - processing entity User
Error: Please provide either 'type' or 'entity' attribute in User.id
    at ReflectMetadataProvider.initPropertyType (/Users/jonah/code/async-up/node_modules/@mikro-orm/core/metadata/ReflectMetadataProvider.js:14:19)
    at /Users/jonah/code/async-up/node_modules/@mikro-orm/core/metadata/ReflectMetadataProvider.js:9:54
    at ReflectMetadataProvider.initProperties (/Users/jonah/code/async-up/node_modules/@mikro-orm/core/metadata/MetadataProvider.js:26:23)
    at ReflectMetadataProvider.loadEntityMetadata (/Users/jonah/code/async-up/node_modules/@mikro-orm/core/metadata/ReflectMetadataProvider.js:9:20)
    at MetadataDiscovery.discoverEntity (/Users/jonah/code/async-up/node_modules/@mikro-orm/core/metadata/MetadataDiscovery.js:174:41)
    at MetadataDiscovery.discoverReferences (/Users/jonah/code/async-up/node_modules/@mikro-orm/core/metadata/MetadataDiscovery.js:123:24)
    at MetadataDiscovery.findEntities (/Users/jonah/code/async-up/node_modules/@mikro-orm/core/metadata/MetadataDiscovery.js:74:20)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at async MetadataDiscovery.discover (/Users/jonah/code/async-up/node_modules/@mikro-orm/core/metadata/MetadataDiscovery.js:34:9)
    at async Function.init (/Users/jonah/code/async-up/node_modules/@mikro-orm/core/MikroORM.js:40:24)
```

**To Reproduce**
1. Create a new NextJS project
2. Create a config file for mikro-orm. Here is mine for example
```
import 'reflect-metadata';
import { TsMorphMetadataProvider } from '@mikro-orm/reflection';
import { User } from "../entities/User";

export default {
  dbName: "async-up",
  type: "mysql",
  host: process.env.MYSQL_HOST,
  port: Number(process.env.MYSQL_PORT),
  user: process.env.MYSQL_USERNAME,
  password: process.env.MYSQL_PASSWORD,
  metadataProvider: TsMorphMetadataProvider,
  entities: [User],
  discovery: { disableDynamicFileAccess: true },
  debug: process.env.NODE_ENV === "development",
};

```
3. Attempt to initialize the ORM inside a JS file in the API.
```
import "reflect-metadata";
import { MikroORM } from "@mikro-orm/core";
import config from "../../../database/mikro-orm";

export default async (req, res) => {
  const orm = await MikroORM.init(config);
};
```

**Expected behavior**
I expected the request to the API to return nothing except a 200 response. Instead I got "Internal Server Error".

In console reads out the error mentioned above.

**Additional context**
I am trying to create an API in NextJS that hits our database for an amount of information small enough I'd prefer to not set up a dedicated API server. I loosely based this off the example repo for serverless functions and mikro-orm.

For context, here is my entity:

```
import 'reflect-metadata';
import {
  PrimaryKey,
  Entity,
  Property,  
  Unique,
} from "@mikro-orm/core";

import { BaseEntity } from "./BaseEntity";

@Entity()
export class User extends BaseEntity {
  @PrimaryKey()
  id!: number;

  @Property()
  name!: string;

  @Property()
  @Unique()
  email: string;

  @Property({ nullable: true })
  age?: number;

  @Property()
  termsAccepted = false;

  @Property({ nullable: true })
  born?: Date;

  constructor(name: string, email: string) {
    super();
    this.name = name;
    this.email = email;
  }
}
```

**Versions**

```
"@mikro-orm/core": "^4.0.7",
"@mikro-orm/mysql": "^4.0.7",
"@mikro-orm/reflection": "^4.0.7",
"@mikro-orm/sql-highlighter": "github:mikro-orm/sql-highlighter",
```

**Comment by B4nan** [comment]
Your config does not make sense, you are talking about reflect-metadata, but you explicitly set that you want ts-morph, as well as `disableDynamicFileAccess` - those are basically contradictions. You can't use ts-morph and `disableDynamicFileAccess` together. 

Also I am quite sure that `import 'reflect-metadata';` does not belong to the entity file - that should be imported only once per the whole app, and as soon as possible, so usually in some bootstrap script like `main.ts` or `app.ts`.

But I would expect different issue -- never tried nextjs myself, but I though they use babel to compile TS - and that requires additional setup to allow using reflect-metadata, as babel do not support `emitDecoratorMetadata` without plugins. Namely this one should be needed:

https://github.com/leonardfactory/babel-plugin-transform-typescript-metadata

I can assure you that the file type where you initialize the ORM has no effect. There are issues about babel here, so be sure to check those.

**Comment by jonahallibone** [comment]
@B4nan Thanks for the quick response. I actually got it working now. I wonder if I could actually create a contrived example of this and add it to your examples repo? I see a lot of people trying to use TypeORM with NextJS which creates a host of issues with HMR and serverless. This was actually my motivation to switch to MikroOrm. Let me know fi you have any thoughts on this.

**Comment by B4nan** [comment]
Yeah agreed, that would be very helpful!

**Comment by jonahallibone** [comment]
@B4nan Here is what I've come up with thus far if you would like to look it over:

https://github.com/jonahallibone/mikro-orm-nextjs

**Comment by B4nan** [comment]
Would be good to verify both dev and prod builds are working, other than it looks good. Would you send a PR adding the link to readme and docs? Like this one #780 (and feel free to add it as submodule to https://github.com/mikro-orm/mikro-orm-examples/ too).

**Comment by jonahallibone** [comment]
@B4nan I'm actually not sure how to add as a submodule. Will research. But I can get the PR for the readme done.

**Comment by B4nan** [comment]
`git submodule add ...url... ...path...`, easy as that.

https://chrisjean.com/git-submodules-adding-using-removing-and-updating/

**Comment by jonahallibone** [comment]
@B4nan I actually just noticed a strange behavior maybe you can advise on... the CLI commands `npx mikro-orm ...` only seem to work when the dev server is running. Is there some configuration piece I'm not seeing? It also started randomly throwing
```
- configuration not found (Cannot find module '@mikro-orm/mysql'
Require stack:
- /Users/jonah/.npm-global/lib/node_modules/@mikro-orm/cli/node_modules/@mikro-orm/core/utils/Configuration.js
- /Users/jonah/.npm-global/lib/node_modules/@mikro-orm/cli/node_modules/@mikro-orm/core/utils/index.js
- /Users/jonah/.npm-global/lib/node_modules/@mikro-orm/cli/node_modules/@mikro-orm/core/types/ArrayType.js
- /Users/jonah/.npm-global/lib/node_modules/@mikro-orm/cli/node_modules/@mikro-orm/core/types/index.js
- /Users/jonah/.npm-global/lib/node_modules/@mikro-orm/cli/node_modules/@mikro-orm/core/metadata/EntitySchema.js
- /Users/jonah/.npm-global/lib/node_modules/@mikro-orm/cli/node_modules/@mikro-orm/core/metadata/index.js
- /Users/jonah/.npm-global/lib/node_modules/@mikro-orm/cli/node_modules/@mikro-orm/core/MikroORM.js
- /Users/jonah/.npm-global/lib/node_modules/@mikro-orm/cli/node_modules/@mikro-orm/core/index.js
- /Users/jonah/.npm-global/lib/node_modules/@mikro-orm/cli/CLIConfigurator.js
- /Users/jonah/.npm-global/lib/node_modules/@mikro-orm/cli/cli.js)
```
as well

**Comment by B4nan** [comment]
Looks like you do not have the cli package installed locally? If you install it globally, you wont have the driver package...

**Comment by jonahallibone** [comment]
@B4nan That was it. Thanks a bunch.

**Comment by jonahallibone** [comment]
@B4nan https://medium.com/@jonahallibone_20420/next-js-9-mikroorm-eb6f6e08e1a1?sk=ede639ec2b25d7691313e88cb0d4526f

**Comment by B4nan** [comment]
Great, thanks for that! Btw you can omit the `discovery: { disableDynamicFileAccess: false },`, its the default, also all this toggle currently does is to trigger additional validation and sets the metadata provider to reflect-metadata (but that is the default provider in v4 anyway).

**Comment by jonahallibone** [comment]
@B4nan I removed the file discovery. I was curious about the migrations in the CLI. I seem to get the same issue everytime I try and create and run a new migration with relationships. Is there some order I should be using the CLI tools in to mitigate this issue?

<img width="1097" alt="Screen Shot 2020-09-28 at 7 26 03 PM" src="https://user-images.githubusercontent.com/8886318/94496019-400f6a00-01c1-11eb-8a8d-b2f983d1ec7b.png">

**Comment by B4nan** [comment]
Feel free to open new issue for that if you can put together a reproduction. But if it is about renaming columns, that is in general very problematic (and can result in error like the above), as mysql won't even allow you to change FK name, so it is basically a won't fix. 

The schema diffing is far away from perfect, sometimes it tries to do things that won't work, sometimes it is not even possible to do things due to db limitations (e.g. postgres won't allow you to change column type if the values cannot be casted to the new type). 

You should always verify the generated migration and fix it if needed.
